#                                Home Credit Default Risk - Modelling

## Completed By  

## Sakshi Pandey

## Table Of Contents
##### 1. Introduction
##### 2. Data Exploration

    2.1 Loading Libraries

    2.2 Loading the Dataset

    2.3 Explore the Dataset

##### 3. Dataset Cleaning And Imputing
##### 4. Clean the Test Dataset
##### 5. Majority Class Of Target Variable
##### 6. Pre Modeling
##### 7. Logistic Regression Model




### 1. Introduction

Home Credit Default Risk - Modelling. In this notebook, we delve into the challenge of credit default risk at Home Credit, a company dedicated to offering loans to individuals with limited or no credit history. This demographic often falls prey to unscrupulous lenders, underscoring the importance of ensuring a fair lending experience.

To tackle this issue, Home Credit harnesses alternative data sources like telecom and transactional data to predict clients' repayment capabilities. The analysis kicks off with data cleaning, preparing the dataset for modeling. Then, we build predictive models using various machine learning techniques, evaluating their effectiveness in predicting credit default risk through metrics like accuracy and ROC values.

Our goal is to pinpoint the most effective model based on Kaggle scores. We start with Logistic Regression and subsequently explore Random Forest, XGBoost, and Light XGBoost to gauge their performance in this regard.

### 2. Data Preparation and Exploration

#### 2.1 Loading required Libraries

In [1]:
import warnings
warnings.filterwarnings('ignore')
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

#### 2.2 Loading Dataset

In [2]:
# Loading the application_train dataset.
data_appTrain=pd.read_csv("application_train.csv")

#### 2.3 Explore Dataset

In [3]:
# show first 5 rows of the dataset.
data_appTrain.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21  \
0  ...                 0                0                0                0   
1  ...                 0                0                0                0   
2  ...                 0                0                0                0   
3  ...                 0                0                0                0   
4  ...                 0                0                0                0   

  AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        NaN                       NaN   
4                        0.0                       0.0   

   AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         NaN                        NaN   
4                         0.0                        0.0   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
0                        0.0                         1.0  
1                        0.0                         0.0  
2                        0.0                         0.0  
3                        NaN                         NaN  
4                        0.0                         0.0  

[5 rows x 122 columns]

In [4]:
# show last 5 rows of the dataset.
data_appTrain.tail()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
307506      456251       0         Cash loans           M            N   
307507      456252       0         Cash loans           F            N   
307508      456253       0         Cash loans           F            N   
307509      456254       1         Cash loans           F            N   
307510      456255       0         Cash loans           F            N   

       FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  \
307506               N             0          157500.0    254700.0   
307507               Y             0           72000.0    269550.0   
307508               Y             0          153000.0    677664.0   
307509               Y             0          171000.0    370107.0   
307510               N             0          157500.0    675000.0   

        AMT_ANNUITY  ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20  \
307506      27558.0  ...                 0                0                0   
307507      12001.5  ...                 0                0                0   
307508      29979.0  ...                 0                0                0   
307509      20205.0  ...                 0                0                0   
307510      49117.5  ...                 0                0                0   

       FLAG_DOCUMENT_21 AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
307506                0                        NaN                       NaN   
307507                0                        NaN                       NaN   
307508                0                        1.0                       0.0   
307509                0                        0.0                       0.0   
307510                0                        0.0                       0.0   

        AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
307506                         NaN                        NaN   
307507                         NaN                        NaN   
307508                         0.0                        1.0   
307509                         0.0                        0.0   
307510                         0.0                        2.0   

        AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
307506                        NaN                         NaN  
307507                        NaN                         NaN  
307508                        0.0                         1.0  
307509                        0.0                         0.0  
307510                        0.0                         1.0  

[5 rows x 122 columns]

By utilizing the head() and tail() functions on the dataset, we observe that it consists of 122 columns. Further using shape function, we can get to know the number of total rows present in the dataset. Among the columns, the significant columns are SK_ID_CURR, representing the loan application ID, and TARGET, which denotes the target variable indicating default.

In [5]:
# Determining the number of rows and columns in the dataset
data_appTrain.shape

(307511, 122)

The shape (307511, 122) of the dataset indicates that the dataset contains 307,511 rows and 122 columns.

In [6]:
data_appTrain.duplicated().sum()

0

All rows are unique as the sum of duplicate rows comes out to be 0.

In [7]:
# Show the column names
data_appTrain.columns

Index(['SK_ID_CURR', 'TARGET', 'NAME_CONTRACT_TYPE', 'CODE_GENDER',
       'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'CNT_CHILDREN', 'AMT_INCOME_TOTAL',
       'AMT_CREDIT', 'AMT_ANNUITY',
       ...
       'FLAG_DOCUMENT_18', 'FLAG_DOCUMENT_19', 'FLAG_DOCUMENT_20',
       'FLAG_DOCUMENT_21', 'AMT_REQ_CREDIT_BUREAU_HOUR',
       'AMT_REQ_CREDIT_BUREAU_DAY', 'AMT_REQ_CREDIT_BUREAU_WEEK',
       'AMT_REQ_CREDIT_BUREAU_MON', 'AMT_REQ_CREDIT_BUREAU_QRT',
       'AMT_REQ_CREDIT_BUREAU_YEAR'],
      dtype='object', length=122)

The output displays the column names of the dataset. It includes various features such as 'SK_ID_CURR' representing the loan application ID, 'TARGET' indicating the target variable for default, 'NAME_CONTRACT_TYPE' specifying the type of loan contract, 'CODE_GENDER' denoting the gender of the applicant, 'FLAG_OWN_CAR' and 'FLAG_OWN_REALTY' indicating whether the applicant owns a car or real estate property respectively, 'CNT_CHILDREN' representing the number of children the applicant has, 'AMT_INCOME_TOTAL' indicating the total income of the applicant, 'AMT_CREDIT' representing the credit amount requested, 'AMT_ANNUITY' denoting the loan annuity, and several other features encompassing financial and personal information. 

In [8]:
# Provide data type of every column in the dataset.
data_appTrain.dtypes

SK_ID_CURR                      int64
TARGET                          int64
NAME_CONTRACT_TYPE             object
CODE_GENDER                    object
FLAG_OWN_CAR                   object
                               ...   
AMT_REQ_CREDIT_BUREAU_DAY     float64
AMT_REQ_CREDIT_BUREAU_WEEK    float64
AMT_REQ_CREDIT_BUREAU_MON     float64
AMT_REQ_CREDIT_BUREAU_QRT     float64
AMT_REQ_CREDIT_BUREAU_YEAR    float64
Length: 122, dtype: object

The dataset comprises columns of integer (int64), floating-point (float64), and object data types.

### 3. Dataset Cleaning And Imputing

In [9]:
#examiinge the numeric cloumns.
numeric_cols = data_appTrain.select_dtypes(exclude=['object']).columns
numeric_summary = data_appTrain[numeric_cols].describe()
numeric_summary

SK_ID_CURR         TARGET   CNT_CHILDREN  AMT_INCOME_TOTAL  \
count  307511.000000  307511.000000  307511.000000      3.075110e+05   
mean   278180.518577       0.080729       0.417052      1.687979e+05   
std    102790.175348       0.272419       0.722121      2.371231e+05   
min    100002.000000       0.000000       0.000000      2.565000e+04   
25%    189145.500000       0.000000       0.000000      1.125000e+05   
50%    278202.000000       0.000000       0.000000      1.471500e+05   
75%    367142.500000       0.000000       1.000000      2.025000e+05   
max    456255.000000       1.000000      19.000000      1.170000e+08   

         AMT_CREDIT    AMT_ANNUITY  AMT_GOODS_PRICE  \
count  3.075110e+05  307499.000000     3.072330e+05   
mean   5.990260e+05   27108.573909     5.383962e+05   
std    4.024908e+05   14493.737315     3.694465e+05   
min    4.500000e+04    1615.500000     4.050000e+04   
25%    2.700000e+05   16524.000000     2.385000e+05   
50%    5.135310e+05   24903.000000     4.500000e+05   
75%    8.086500e+05   34596.000000     6.795000e+05   
max    4.050000e+06  258025.500000     4.050000e+06   

       REGION_POPULATION_RELATIVE     DAYS_BIRTH  DAYS_EMPLOYED  ...  \
count               307511.000000  307511.000000  307511.000000  ...   
mean                     0.020868  -16036.995067   63815.045904  ...   
std                      0.013831    4363.988632  141275.766519  ...   
min                      0.000290  -25229.000000  -17912.000000  ...   
25%                      0.010006  -19682.000000   -2760.000000  ...   
50%                      0.018850  -15750.000000   -1213.000000  ...   
75%                      0.028663  -12413.000000    -289.000000  ...   
max                      0.072508   -7489.000000  365243.000000  ...   

       FLAG_DOCUMENT_18  FLAG_DOCUMENT_19  FLAG_DOCUMENT_20  FLAG_DOCUMENT_21  \
count     307511.000000     307511.000000     307511.000000     307511.000000   
mean           0.008130          0.000595          0.000507          0.000335   
std            0.089798          0.024387          0.022518          0.018299   
min            0.000000          0.000000          0.000000          0.000000   
25%            0.000000          0.000000          0.000000          0.000000   
50%            0.000000          0.000000          0.000000          0.000000   
75%            0.000000          0.000000          0.000000          0.000000   
max            1.000000          1.000000          1.000000          1.000000   

       AMT_REQ_CREDIT_BUREAU_HOUR  AMT_REQ_CREDIT_BUREAU_DAY  \
count               265992.000000              265992.000000   
mean                     0.006402                   0.007000   
std                      0.083849                   0.110757   
min                      0.000000                   0.000000   
25%                      0.000000                   0.000000   
50%                      0.000000                   0.000000   
75%                      0.000000                   0.000000   
max                      4.000000                   9.000000   

       AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
count               265992.000000              265992.000000   
mean                     0.034362                   0.267395   
std                      0.204685                   0.916002   
min                      0.000000                   0.000000   
25%                      0.000000                   0.000000   
50%                      0.000000                   0.000000   
75%                      0.000000                   0.000000   
max                      8.000000                  27.000000   

       AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
count              265992.000000               265992.000000  
mean                    0.265474                    1.899974  
std                     0.794056                    1.869295  
min                     0.000000                    0.000000  
25%              

In examining the dataset, several outliers have been identified across various numeric columns. Notably, the 'CNT_CHILDREN' column exhibits a maximum value of 19, which exceeds typical family sizes and may warrant further investigation. Similarly, the 'AMT_INCOME_TOTAL' column displays a maximum value of 117,000,000, significantly higher than the 75th percentile and mean values, indicating a potential outlier. Another striking observation is in the 'DAYS_EMPLOYED' column, where the maximum value equates to approximately 1000 years, a clear anomaly suggesting data entry errors or unconventional coding. These outliers could significantly impact analyses and model performance if left unaddressed, highlighting the importance of outlier detection and appropriate handling techniques in data preprocessing.

In [10]:
# examine the categorical cloumns.
categorical_cols = data_appTrain.select_dtypes(include=['object']).columns
for col in categorical_cols:
    print(f"\nUnique values in {col}:")
    print(data_appTrain[col].value_counts())


Unique values in NAME_CONTRACT_TYPE:
NAME_CONTRACT_TYPE
Cash loans         278232
Revolving loans     29279
Name: count, dtype: int64

Unique values in CODE_GENDER:
CODE_GENDER
F      202448
M      105059
XNA         4
Name: count, dtype: int64

Unique values in FLAG_OWN_CAR:
FLAG_OWN_CAR
N    202924
Y    104587
Name: count, dtype: int64

Unique values in FLAG_OWN_REALTY:
FLAG_OWN_REALTY
Y    213312
N     94199
Name: count, dtype: int64

Unique values in NAME_TYPE_SUITE:
NAME_TYPE_SUITE
Unaccompanied      248526
Family              40149
Spouse, partner     11370
Children             3267
Other_B              1770
Other_A               866
Group of people       271
Name: count, dtype: int64

Unique values in NAME_INCOME_TYPE:
NAME_INCOME_TYPE
Working                 158774
Commercial associate     71617
Pensioner                55362
State servant            21703
Unemployed                  22
Student                     18
Businessman                 10
Maternity leave              

The output generates the unique values and their respective counts for each categorical column in the dataset. For instance, in the 'NAME_CONTRACT_TYPE' column, there are primarily two types: 'Cash loans' and 'Revolving loans', with 'Cash loans' being significantly more prevalent. Similarly, the 'CODE_GENDER' column contains three categories: 'F' (female), 'M' (male), and 'XNA', with 'F' being the most frequent gender category. Notably, 'XNA' may indicate missing or undefined data. Each categorical column showcases the diversity within the dataset, offering insights into the distribution and prevalence of different categories across various features.

#### Calculate the percentage of missing values

In [37]:
missing_data = data_appTrain.isna().sum() / len(data_appTrain)
missing_data_sorted = missing_data.sort_values(ascending=False)
missing_data_sorted

FLOORSMAX_AVG                   0.497608
FLOORSMAX_MODE                  0.497608
FLOORSMAX_MEDI                  0.497608
YEARS_BEGINEXPLUATATION_AVG     0.487810
YEARS_BEGINEXPLUATATION_MODE    0.487810
                                  ...   
NAME_HOUSING_TYPE               0.000000
NAME_FAMILY_STATUS              0.000000
NAME_EDUCATION_TYPE             0.000000
NAME_INCOME_TYPE                0.000000
SK_ID_CURR                      0.000000
Length: 80, dtype: float64

The dataset exhibits varying degrees of missing data across its columns. Notably, features like COMMONAREA_MEDI, COMMONAREA_AVG, and COMMONAREA_MODE have approximately 70% missing values each, suggesting potential limited informativeness due to the substantial absence of data. Conversely, certain columns, such as NAME_HOUSING_TYPE, NAME_FAMILY_STATUS, NAME_EDUCATION_TYPE, NAME_INCOME_TYPE, and SK_ID_CURR, show no missing values. This discrepancy underscores the importance of careful consideration when utilizing features with high proportions of missing data, as they may not provide reliable insights. Conversely, columns devoid of missing values offer robust potential for analysis and modeling. We will drop the columns that have more than 50% of the missing values and impute the columns with mean having less than 50% of the missing values.

In [38]:
# Select numeric columns
col_numeric= data_appTrain.select_dtypes(include=['number'])
missings_col=col_numeric.loc[:,col_numeric.isna().mean() >= 0.50]
highly_missing_features = missings_col.columns
missings_col.columns

# Drop columns with more than 50% missing values
data_appTrain =data_appTrain.drop(highly_missing_features,axis=1)
data_appTrain.head()


SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21  \
0  ...                 0                0                0                0   
1  ...                 0                0                0                0   
2  ...                 0                0                0                0   
3  ...                 0                0                0                0   
4  ...                 0                0                0                0   

  AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        NaN                       NaN   
4                        0.0                       0.0   

   AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         NaN                        NaN   
4                         0.0                        0.0   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
0                        0.0                         1.0  
1                        0.0                         0.0  
2                        0.0                         0.0  
3                        NaN                         NaN  
4                        0.0                         0.0  

[5 rows x 80 columns]

After dropping numeric columns, we can see that we have 80 columns.

In [39]:
#dropping object columns
application_train_obj= data_appTrain.select_dtypes(include=['object'])
missings_objCol=application_train_obj.loc[:,application_train_obj.isna().mean() >= 0.50]
highly_missing_features_obj = missings_objCol.columns
missings_objCol.columns

#drop object columns as found above
data_appTrain=data_appTrain.drop(highly_missing_features_obj,axis="columns")
data_appTrain.head()


SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21  \
0  ...                 0                0                0                0   
1  ...                 0                0                0                0   
2  ...                 0                0                0                0   
3  ...                 0                0                0                0   
4  ...                 0                0                0                0   

  AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        NaN                       NaN   
4                        0.0                       0.0   

   AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         NaN                        NaN   
4                         0.0                        0.0   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
0                        0.0                         1.0  
1                        0.0                         0.0  
2                        0.0                         0.0  
3                        NaN                         NaN  
4                        0.0                         0.0  

[5 rows x 80 columns]

We can see that after dropping the numeric and object type columns we finally have 80 columns which means, in object type columns, there is no missing values above 50%.

#### Imputing Null values

In [41]:
#Imputing obejct columns with mode and numeric with mean
for column in data_appTrain.columns:
    if data_appTrain[column].dtype == 'object':
        data_appTrain[column] = data_appTrain[column].fillna(data_appTrain[column].mode().iloc[0])
    else:
        data_appTrain[column] = data_appTrain[column].fillna(data_appTrain[column].mean())


In [42]:
missing_data=data_appTrain.isna().sum()
missing_data.sort_values(ascending=False)

SK_ID_CURR                    0
TARGET                        0
FLAG_DOCUMENT_6               0
FLAG_DOCUMENT_5               0
FLAG_DOCUMENT_4               0
                             ..
FLAG_PHONE                    0
FLAG_CONT_MOBILE              0
FLAG_WORK_PHONE               0
FLAG_EMP_PHONE                0
AMT_REQ_CREDIT_BUREAU_YEAR    0
Length: 80, dtype: int64

The output shows that there are no missing values (nulls) in any of the columns in the DataFrame. Each column has a count of 0 null values, indicating that the DataFrame is now free of missing data.

#### Remove outliers

In [43]:
def remove_outliers(col):
  Q1,Q3=col.quantile([.25,.75])
  IQR=Q3-Q1
  lower_range=Q1-(1.5*IQR)
  upper_range=Q3+(1.5*IQR)
  return lower_range,upper_range

In [45]:
for i in data_appTrain.columns:
 if data_appTrain[i].dtype != 'object':
    lowlevel,uplevel=remove_outliers(data_appTrain['DAYS_EMPLOYED'])
data_appTrain['DAYS_EMPLOYED']=np.where(data_appTrain['DAYS_EMPLOYED']>uplevel,uplevel,data_appTrain['DAYS_EMPLOYED'])
data_appTrain['DAYS_EMPLOYED']=np.where(data_appTrain['DAYS_EMPLOYED']<lowlevel,lowlevel,data_appTrain['DAYS_EMPLOYED'])


In [47]:
data_appTrain['DAYS_EMPLOYED'].describe()

count    307511.000000
mean      -1203.542428
std        2732.404969
min       -6466.500000
25%       -2760.000000
50%       -1213.000000
75%        -289.000000
max        3417.500000
Name: DAYS_EMPLOYED, dtype: float64

After eliminating the outliers, the highest value observed for the DAY_EMPLOYED variable is 3417 days.

In [48]:
# Flag document just flags the data whether customer submitted document or not so better to remove

flagged_documents = ['WEEKDAY_APPR_PROCESS_START','FLAG_DOCUMENT_2','FLAG_DOCUMENT_3','FLAG_DOCUMENT_4','FLAG_DOCUMENT_5','FLAG_DOCUMENT_6','FLAG_DOCUMENT_7',
'FLAG_DOCUMENT_8','FLAG_DOCUMENT_9','FLAG_DOCUMENT_10','FLAG_DOCUMENT_11','FLAG_DOCUMENT_12','FLAG_DOCUMENT_13','FLAG_DOCUMENT_14',
'FLAG_DOCUMENT_15','FLAG_DOCUMENT_16','FLAG_DOCUMENT_17','FLAG_DOCUMENT_18','FLAG_DOCUMENT_19','FLAG_DOCUMENT_20','FLAG_DOCUMENT_21']

In [49]:
data_appTrain.drop(columns=flagged_documents, inplace=True)

In [50]:
data_appTrain.shape

(307511, 59)

In [51]:
data_appTrain.to_csv(r'train.csv', index=False)

We're removing the columns related to document flags from the dataset, which reduces the total number of columns to 52. The updated dataset, aimed at enhancing its usability for future analysis or modeling, is saved as a CSV file named 'train.csv'. This step streamlines the dataset by eliminating unnecessary or duplicate columns.

### 4. Cleaning Test Dataset

In [52]:
#load test dataset
test_data=pd.read_csv("application_test.csv")

In [53]:
test_data.head() 

SK_ID_CURR NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  \
0      100001         Cash loans           F            N               Y   
1      100005         Cash loans           M            N               Y   
2      100013         Cash loans           M            Y               Y   
3      100028         Cash loans           F            N               Y   
4      100038         Cash loans           M            Y               N   

   CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  AMT_GOODS_PRICE  \
0             0          135000.0    568800.0      20560.5         450000.0   
1             0           99000.0    222768.0      17370.0         180000.0   
2             0          202500.0    663264.0      69777.0         630000.0   
3             2          315000.0   1575000.0      49018.5        1575000.0   
4             1          180000.0    625500.0      32067.0         625500.0   

   ... FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21  \
0  ...                0                0                0                0   
1  ...                0                0                0                0   
2  ...                0                0                0                0   
3  ...                0                0                0                0   
4  ...                0                0                0                0   

  AMT_REQ_CREDIT_BUREAU_HOUR  AMT_REQ_CREDIT_BUREAU_DAY  \
0                        0.0                        0.0   
1                        0.0                        0.0   
2                        0.0                        0.0   
3                        0.0                        0.0   
4                        NaN                        NaN   

   AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         0.0                        0.0   
4                         NaN                        NaN   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
0                        0.0                         0.0  
1                        0.0                         3.0  
2                        1.0                         4.0  
3                        0.0                         3.0  
4                        NaN                         NaN  

[5 rows x 121 columns]

In [54]:
test_data.tail() 

SK_ID_CURR NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  \
48739      456221         Cash loans           F            N               Y   
48740      456222         Cash loans           F            N               N   
48741      456223         Cash loans           F            Y               Y   
48742      456224         Cash loans           M            N               N   
48743      456250         Cash loans           F            Y               N   

       CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
48739             0          121500.0    412560.0      17473.5   
48740             2          157500.0    622413.0      31909.5   
48741             1          202500.0    315000.0      33205.5   
48742             0          225000.0    450000.0      25128.0   
48743             0          135000.0    312768.0      24709.5   

       AMT_GOODS_PRICE  ... FLAG_DOCUMENT_18 FLAG_DOCUMENT_19  \
48739         270000.0  ...                0                0   
48740         495000.0  ...                0                0   
48741         315000.0  ...                0                0   
48742         450000.0  ...                0                0   
48743         270000.0  ...                0                0   

      FLAG_DOCUMENT_20 FLAG_DOCUMENT_21 AMT_REQ_CREDIT_BUREAU_HOUR  \
48739                0                0                        0.0   
48740                0                0                        NaN   
48741                0                0                        0.0   
48742                0                0                        0.0   
48743                0                0                        0.0   

       AMT_REQ_CREDIT_BUREAU_DAY  AMT_REQ_CREDIT_BUREAU_WEEK  \
48739                        0.0                         0.0   
48740                        NaN                         NaN   
48741                        0.0                         0.0   
48742                        0.0                         0.0   
48743                        0.0                         0.0   

       AMT_REQ_CREDIT_BUREAU_MON  AMT_REQ_CREDIT_BUREAU_QRT  \
48739                        0.0                        0.0   
48740                        NaN                        NaN   
48741                        0.0                        3.0   
48742                        0.0                        0.0   
48743                        0.0                        1.0   

       AMT_REQ_CREDIT_BUREAU_YEAR  
48739                         1.0  
48740                         NaN  
48741                         1.0  
48742                         2.0  
48743                         4.0  

[5 rows x 121 columns]

There are 121 columns in the dataset.

In [55]:
test_data.duplicated().sum()

0

There are no duplicate columns in the test dataset

In [56]:
print(test_data.shape)
test_data.info()

(48744, 121)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48744 entries, 0 to 48743
Columns: 121 entries, SK_ID_CURR to AMT_REQ_CREDIT_BUREAU_YEAR
dtypes: float64(65), int64(40), object(16)
memory usage: 45.0+ MB


The dataset comprises 48,744 entries organized into 121 columns. These entries are indexed from 0 to 48,743. The columns consist of various data types: 65 columns are of float type, 40 columns are of integer type, and 16 columns are of object type. 

Let's exclude all the columns that were removed from the train dataset.

In [68]:
col_numeric= test_data.select_dtypes(include=['number'])
missings_col_test=col_numeric.loc[:,col_numeric.isna().mean() >= 0.50]
highly_missing_features_test = missings_col_test.columns
missings_col_test.columns
test_data_clean=test_data.drop(highly_missing_features_test,axis=1) # Dropping the numeric columns.
test_data_clean.head()

SK_ID_CURR NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  \
0      100001         Cash loans           F            N               Y   
1      100005         Cash loans           M            N               Y   
2      100013         Cash loans           M            Y               Y   
3      100028         Cash loans           F            N               Y   
4      100038         Cash loans           M            Y               N   

   CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  AMT_GOODS_PRICE  \
0             0          135000.0    568800.0      20560.5         450000.0   
1             0           99000.0    222768.0      17370.0         180000.0   
2             0          202500.0    663264.0      69777.0         630000.0   
3             2          315000.0   1575000.0      49018.5        1575000.0   
4             1          180000.0    625500.0      32067.0         625500.0   

   ... FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21  \
0  ...                0                0                0                0   
1  ...                0                0                0                0   
2  ...                0                0                0                0   
3  ...                0                0                0                0   
4  ...                0                0                0                0   

  AMT_REQ_CREDIT_BUREAU_HOUR  AMT_REQ_CREDIT_BUREAU_DAY  \
0                        0.0                        0.0   
1                        0.0                        0.0   
2                        0.0                        0.0   
3                        0.0                        0.0   
4                        NaN                        NaN   

   AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         0.0                        0.0   
4                         NaN                        NaN   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
0                        0.0                         0.0  
1                        0.0                         3.0  
2                        1.0                         4.0  
3                        0.0                         3.0  
4                        NaN                         NaN  

[5 rows x 93 columns]

In [69]:
application_test_obj= test_data.select_dtypes(include=['object'])
missings_objColtest=application_test_obj.loc[:,application_test_obj.isna().mean() >= 0.50]
highly_missing_features_obj = missings_objColtest.columns
missings_objColtest.columns
test_data_clean=test_data_clean.drop(highly_missing_features_obj,axis=1) # Dropping the object columns.
test_data_clean.head()

SK_ID_CURR NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  \
0      100001         Cash loans           F            N               Y   
1      100005         Cash loans           M            N               Y   
2      100013         Cash loans           M            Y               Y   
3      100028         Cash loans           F            N               Y   
4      100038         Cash loans           M            Y               N   

   CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  AMT_GOODS_PRICE  \
0             0          135000.0    568800.0      20560.5         450000.0   
1             0           99000.0    222768.0      17370.0         180000.0   
2             0          202500.0    663264.0      69777.0         630000.0   
3             2          315000.0   1575000.0      49018.5        1575000.0   
4             1          180000.0    625500.0      32067.0         625500.0   

   ... FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21  \
0  ...                0                0                0                0   
1  ...                0                0                0                0   
2  ...                0                0                0                0   
3  ...                0                0                0                0   
4  ...                0                0                0                0   

  AMT_REQ_CREDIT_BUREAU_HOUR  AMT_REQ_CREDIT_BUREAU_DAY  \
0                        0.0                        0.0   
1                        0.0                        0.0   
2                        0.0                        0.0   
3                        0.0                        0.0   
4                        NaN                        NaN   

   AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         0.0                        0.0   
4                         NaN                        NaN   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
0                        0.0                         0.0  
1                        0.0                         3.0  
2                        1.0                         4.0  
3                        0.0                         3.0  
4                        NaN                         NaN  

[5 rows x 92 columns]

So, we have 93 rows after dropping numeric columns with missing values greater than 50% and 92 columns after dropping object columns

In [70]:
missingdata=(test_data_clean.isna().sum()/len(test_data_clean))
missingdata.sort_values(ascending=False) # Find the percentage of missong values in the other columns in descending order.

WALLSMATERIAL_MODE     0.490173
APARTMENTS_AVG         0.490050
APARTMENTS_MODE        0.490050
APARTMENTS_MEDI        0.490050
HOUSETYPE_MODE         0.484552
                         ...   
NAME_HOUSING_TYPE      0.000000
NAME_FAMILY_STATUS     0.000000
NAME_EDUCATION_TYPE    0.000000
NAME_INCOME_TYPE       0.000000
SK_ID_CURR             0.000000
Length: 92, dtype: float64

The output highlights varying degrees of missing data across columns in the test dataset. Notably, certain features, such as WALLSMATERIAL_MODE and APARTMENTS_AVG, exhibit a high proportion of missing values, around 49%. Conversely, columns like NAME_HOUSING_TYPE and NAME_FAMILY_STATUS show complete data, indicating robustness in these aspects. Managing missing data is essential for maintaining the integrity of analyses and models, prompting the need for strategies like imputation or careful feature selection to address data gaps effectively.

In [71]:
for column in test_data_clean.columns:
    if test_data_clean[column].dtype == 'object':
        test_data_clean[column] = test_data_clean[column].fillna(test_data_clean[column].mode().iloc[0]) # Impute with mode for object columns.
    else:
        test_data_clean[column] = test_data_clean[column].fillna(test_data_clean[column].median())# Impute with median for numeric columns.

In [72]:
missingdata=(test_data_clean.isna().sum()/len(test_data_clean))  # Let's see any missing values are present.
missingdata.sort_values(ascending=False)

SK_ID_CURR                    0.0
TOTALAREA_MODE                0.0
FLAG_DOCUMENT_3               0.0
FLAG_DOCUMENT_2               0.0
DAYS_LAST_PHONE_CHANGE        0.0
                             ... 
REGION_RATING_CLIENT          0.0
CNT_FAM_MEMBERS               0.0
OCCUPATION_TYPE               0.0
FLAG_EMAIL                    0.0
AMT_REQ_CREDIT_BUREAU_YEAR    0.0
Length: 92, dtype: float64

We can see after imputing there are no missing values.

In [73]:
test_data_clean.drop(columns=flagged_documents, inplace=True)

In [74]:
test_data_clean.shape

(48744, 71)

In [75]:
test_data_clean.to_csv(r'test.csv', index=False)

we remove the columns associated with document flags from the testdata_clean dataset, resulting in a refined dataset with 71 columns and 48,744 rows. The modified dataset is saved as 'test.csv', which could be beneficial for subsequent analysis.

### 5. Majority Class Of Target Variable

In [142]:
# The count of entries in the TARGET column is grouped into two categories: those with defaults and those without defaults.
target = data_appTrain.TARGThe count of entries in the TARGET column is grouped into two categories: those with defaults and those without defaults.ET.value_counts()
target

TARGET
0    282686
1     24825
Name: count, dtype: int64

As we see that 0 (no-default) has 282686 where as 1 (default) has 24825. This is an perfect example of the class imbalance as majority of the target variable has only one class. There are far more loans that were repaid on time than loans that were not repaid.This may affect the model and results in biased predictions toward the non default

In [143]:
# Calculating the percentage of target class.
percenttarget=(target.values/len(data_app)*100)
percenttarget

array([91.92711805,  8.07288195])

We can see that 0 (no-default) has 91.927% where as 1 (default) has 8.072% and the majority class classifier has 91.927% accuracy for the 0 (no-default) classifier.

### 6. Pre-Modeling

Loading the libraries required for the modeling.

In [144]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report
import xgboost as xgb
from sklearn.metrics import roc_auc_score
from xgboost import XGBClassifier
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import accuracy_score, mean_squared_error, confusion_matrix

Loading the cleaned train and cleaned test datasets that we cleaned before.

In [145]:
train_data=pd.read_csv("train.csv")
test_data=pd.read_csv("test.csv")
test_id = test_data['SK_ID_CURR']

In [146]:
# Creating object columns
target=train_data.iloc[:,1]
train_data=train_data.drop(columns=["TARGET"])
train_df=train_data.select_dtypes(include=['object'])
test_df=test_data.select_dtypes(include=['object'])

Target variable from the train data is stored in target.
Removing the target variable from the train data and storing into train_data.
Taking only object columns from the train_data and storing it in train_df.
Taking only object columns from the test_data and storing it in test_df.

In [147]:
train_df.shape

(307511, 11)

The train_df has the shape of 11 columns which are only object type.

In [148]:
test_df.shape

(48744, 11)

The test_df has the shape of 11 columns which are only object type.

In [149]:
# Combine train and test data
combined_df = pd.concat([train_df, test_df])

# Initialize the OneHotEncoder with handle_unknown='ignore'
encoder = OneHotEncoder(handle_unknown='ignore')


# Fit and transform the encoder on the combined data
encoded_data_combined_sparse = encoder.fit_transform(combined_df)

encoded_data_combined = encoded_data_combined_sparse.toarray()
# Split the encoded data back into train and test datasets
encoded_df_train = pd.DataFrame(encoded_data_combined[:len(train_df)], columns=encoder.get_feature_names_out(train_df.columns))
encoded_df_test = pd.DataFrame(encoded_data_combined[len(train_df):], columns=encoder.get_feature_names_out(test_df.columns))

Here, we encode the object columns from the train and test dataset to one hot encoding and get dummies for every object columns.

In [150]:
print(encoded_df_train.shape)
encoded_df_test.shape

(307511, 117)


(48744, 117)

The encoded_df_train and encoded_df_test has 117 columns which are one hot encodings for the object columns.

In [151]:
# Dropping the object columns
X_train=train_data.drop(columns=train_df.columns)

Dropping the object columns that are present in the train_data and storing this new dataframe to X_train so that we can add new hot encoding columns.

In [152]:
# Concatenating X_train with encoded dataframe
X_train = pd.concat([X_train, encoded_df_train], axis=1)

Here, we are concating the numeric columns from X_train with one hot encodings to form X_train training dataset.

In [153]:
# Dropping the object columns
X_test=test_data.drop(columns=test_df.columns)

Dropping the object columns that are present in the test_data and storing this new dataframe to X_test so that we can add new hot encoding columns.

In [154]:
# Concatenating X_test with encoded dataframe
X_test = pd.concat([X_test, encoded_df_test], axis=1)

Here, we are concating the numeric columns from X_test with one hot encodings to form X_test testing dataset.

In [155]:
print(X_train.shape)
print(X_test.shape)

(307511, 157)
(48744, 157)


Here, the shape of the X_train and X_test has 157 columns total for the final train and test datasets.

In [156]:
# Imblearn for the undersampling of the class variable.
from imblearn.under_sampling import RandomUnderSampler
under_sampler = RandomUnderSampler(random_state=0)
X_train, target_resampled = under_sampler.fit_resample(X_train,target)
X_train["TARGET"] = target_resampled

Since, this is an class imbalance problem we need to do undersampling from the imblearn library under sampling method to get the same class values.

In [157]:
X_train["TARGET"] .value_counts()

TARGET
0    24825
1    24825
Name: count, dtype: int64

Here we can see that the class imbalance is balanced with 0's 24825 rows and 1's 24825.

In [158]:
X_train.drop('SK_ID_CURR', axis=1, inplace=True)
X_test.drop('SK_ID_CURR', axis=1, inplace=True)

Dropping the SK_ID_CURR column from both X_train and X_test.

In [159]:
Y_train=X_train[['TARGET']]
X_train.drop(columns=["TARGET"],inplace=True)

Dropping TARGET from X_train and saving TARGET variable to Y_train.

In [160]:
print(X_train.shape)
Y_train.shape

(49650, 156)


(49650, 1)

X_train (Traing dataset) has 156 columns and Y_train (Training dataset target variable) has 1 column which is TARGET variable. Both X_train and Y_train has 49650 rows.

In [161]:
print(X_test.shape)

(48744, 156)


The X_test (test dataset) has 156 columns with 48744 rows

In [162]:
# Split the data into training and testing sets
X_train, X_valid, y_train, y_valid = train_test_split(X_train, Y_train, test_size=0.2, random_state=42)

Cross_validation of the X_train (training dataset) into X_Train and X_valid with 80%-20% respectively for the validation of the model and to know about overfitting. Also same way for the Y_train (TARGET) variable into 80-20% split for cross validation.

### 7. Logistic Regression Model

Logistic regression is a statistical model used for binary classification. It calculates the probability of an observation belonging to a particular class based on input features and applies a logistic function to map the output in range between 0 and 1. The class with the highest probability is predicted as the outcome.

Importing the preprocessing from sklearn to preprocess the X_train, X_test and X_valid to center and scale for the logistic model to X_scaled, X_test_scaled and Y_valid_scaled.

In [163]:
# Importing the preprocessing methods from sklearn and transforming the datasets.
from sklearn import preprocessing
scaler = preprocessing.StandardScaler().fit(X_train)
X_scaled = scaler.transform(X_train)
scaler = preprocessing.StandardScaler().fit(X_test)
X_test_scaled = scaler.transform(X_test)
scaler = preprocessing.StandardScaler().fit(X_valid)
X_valid_scaled = scaler.transform(X_valid)

Fitting the logistic regression model from sklearn and fit the model on X_scaled and y_train.

In [164]:
# Fitting the logistic regression model on train dataset.
model = LogisticRegression()
model.fit(X_scaled, y_train)

LogisticRegression()

Predicting the Fitted model on the validation dataset (X_valid) to measure the performance the model.

In [165]:
# Predict the fitted model on the validation dataset.
y_pred = model.predict(X_valid_scaled)

Calculating accuracy and RMSE and AUC (Area Under the Curve)  along with confusion matrix for the logistic regression model with validation dataset.

In [166]:
# Calculate accuracy
accuracy = accuracy_score(y_valid, y_pred)
print(f"Accuracy: {accuracy:.2f}")

# Calculate RMSE
rmse = np.sqrt(mean_squared_error(y_valid, y_pred))
print(f"RMSE: {rmse:.2f}")

# Calculate the ROC-AUC score on the test set
roc_auc = roc_auc_score(y_valid, y_pred)
print(f'ROC-AUC Score: {roc_auc}')

# Generate the confusion matrix
conf_matrix = confusion_matrix(y_valid, y_pred)
print("Confusion Matrix:")
print(conf_matrix)

TP = conf_matrix[1, 1]  # True Positives
TN = conf_matrix[0, 0]  # True Negatives

print(f'True Positives: {TP}')
print(f'True Negatives: {TN}')

FN = conf_matrix[1, 0]  # False Negatives
FP = conf_matrix[0, 1]  # False Positives

print(f'False Negatives: {FN}')
print(f'False Positives: {FP}')
from sklearn.metrics import precision_score, recall_score

# Calculate sentivity
sensitivity = recall_score(y_valid, y_pred)

# Calculate precision
precision = precision_score(y_valid, y_pred)

# Calculate specificity
specificity = TN / (TN + FP)

print(f'Sensitivity (Recall): {sensitivity}')
print(f'Precision: {precision}')
print(f'Specificity: {specificity}')

Accuracy: 0.69
RMSE: 0.56
ROC-AUC Score: 0.6854037354386203
Confusion Matrix:
[[3407 1546]
 [1578 3399]]
True Positives: 3399
True Negatives: 3407
False Negatives: 1578
False Positives: 1546
Sensitivity (Recall): 0.6829415310427969
Precision: 0.6873609706774519
Specificity: 0.6878659398344438


Here the accuracy for validation dataset for logistic model is 0.66.

AUC (Area Under the Curve) for the validation set for logistic model is 0.657.

RMSE for the validation set for logistic model is 0.59

Classification matrix for the validation set for logistic model is

True Positives: 3135
True Negatives: 3394
False Negatives: 1842
False Positives: 1559

Here the sensitivity is = 0.6825.

Precision for the validation dataset is = 0.68737.

Specificity for the validation dataset is = 0.68524.

The model is classifying good for both positive and negative classification around 0.68.


Predicting the model on test data and uploading the output to Kaggle to get kaggle score.

In [167]:
# Predicting the model on test data.
y_pred = model.predict(X_test_scaled)

In [168]:
final_df = pd.DataFrame()
final_df["SK_ID_CURR"] = test_id
final_df["TARGET"] = y_pred
final_df

SK_ID_CURR  TARGET
0          100001       1
1          100005       1
2          100013       0
3          100028       0
4          100038       1
...           ...     ...
48739      456221       0
48740      456222       0
48741      456223       0
48742      456224       0
48743      456250       1

[48744 rows x 2 columns]

In [169]:
final_df.to_csv("logistic.csv", index = False)

The Logistic Regression model achieved a Kaggle score of 0.67383, which is relatively low. Given this suboptimal performance, it is necessary to explore alternative models.